doc/source/_ext/suppress.py

```diff
@@ -0,0 +1,291 @@
+
+.. confval:: suppress_protect = []
+
+    A list of module names that are known to contain a Sphinx logger but
+    that will never be suppressed automatically. This is typically useful
+    when an extension contains submodules declaring loggers which, when
+    imported, result in undesirable side-effects.
+
+.. confval:: suppress_records = []
+
+    A list of message patterns to suppress, possibly filtered by logger.
+
+    .. code-block::
+
+        # suppress messages matching 'error: .*' and 'warning: .*'
+        suppress_records = ['error: .*', 'warning: .*']
+
+        # suppress messages issued by 'sphinx.ext.intersphinx'
+        suppress_records = [('sphinx.ext.intersphinx', '.*')]
+"""
+
+__all__ = ()
+
+import abc
+import importlib
+import inspect
+import itertools
+import logging
+import pkgutil
+import re
+import warnings
+from typing import TYPE_CHECKING
+
+from sphinx.util.logging import NAMESPACE, SphinxLoggerAdapter
+
+if TYPE_CHECKING:
+    from sphinx.application import Sphinx
+    from sphinx.config import Config
+    from sphinx.extension import Extension
+
+
+def partition(predicate, iterable):
+    """
+    Return a pair `(no, yes)`, where *yes* and *no* are subsets of *iterable*
+    over which *predicate* evaluates to |False| and |True| respectively.
+    """
+
+    no, yes = itertools.tee(iterable)
+    no, yes = itertools.filterfalse(predicate, no), filter(predicate, yes)
+    return no, yes
+
+
+def notnone(value):
+    return value is not None
+
+
+ALL = object()
+
+
+def _normalize_level(level):
+    if isinstance(level, int):
+        return level
+
+    # pylint: disable-next=W0212
+    result = logging._nameToLevel.get(level)
+    if result is not None:
+        return result
+
+    return None  # unknown level
+
+
+def _parse_levels(levels):
+    if not isinstance(levels, (list, tuple)):
+        if isinstance(levels, (int, str)):
+            levels = [levels]
+    return list(filter(notnone, map(_normalize_level, levels)))
+
+
+class SphinxSuppressFilter(logging.Filter, metaclass=abc.ABCMeta):
+    def filter(self, record):
+        # type: (logging.LogRecord) -> bool
+        return not self.suppressed(record)
+
+    @abc.abstractmethod
+    def suppressed(self, record):
+        # type: (logging.LogRecord) -> bool
+        pass
+
+
+class SphinxSuppressLogger(SphinxSuppressFilter):
+    r"""A filter suppressing logging records issued by a Sphinx logger."""
+
+    def __init__(self, name, levels=()):
+        """
+        Construct a :class:`SphinxSuppressLogger`.
+
+        :param name: The (real) logger name to suppress.
+        :type name: str
+        :param levels: Optional logging levels to suppress.
+        :type levels: bool | list[int]
+        """
+
+        super().__init__(name)
+        if isinstance(levels, bool):
+            self.levels = ALL if levels else []
+        else:
+            self.levels = _parse_levels(levels)
+
+    def suppressed(self, record):
+        return (
```

Comment:
For logging filters, `filter()` returns true iff the record should be logged. Since I wrote filter as the negation of suppressed, then the code should be changed as follows:

```python
def suppressed(self, record):
    should_log = logging.Filter.filter(self, record)
    return not should_log or self.levels is ALL or record.levelno in self.levels
```